### PR TITLE
Extra performance improvements for book count.

### PR DESCRIPTION
**Pull request by AledWEvans**
Extra performance improvements as casing prevents indexing.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 1]
```diff
@@ -104,7 +104,7 @@ interface VisitRepository : JpaRepository<Visit, Long>, JpaSpecificationExecutor
       "JOIN session_template st ON st.reference = v.session_template_reference " +
       "WHERE p.code = :prisonCode AND " +
       "v.session_template_reference = :sessionTemplateReference AND " +
-      "Date(v.visit_start) = :sessionDate AND " +
+      "(v.visit_start >= :sessionDate AND v.visit_start < CAST(:sessionDate AS DATE) + CAST('1 day' AS INTERVAL)) AND " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:107]: understand more after the call we had. Just a minor comment could we put can this statement in brackets - CAST(:sessionDate AS DATE) + CAST(‘1 day’ AS INTERVAL)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 2]
```diff
@@ -125,7 +125,7 @@ interface VisitRepository : JpaRepository<Visit, Long>, JpaSpecificationExecutor
       "JOIN prison p ON p.id = v.prison_id " +
       "JOIN session_template st ON st.reference = v.session_template_reference " +
       "WHERE p.code = :prisonCode AND " +
-      "Date(v.visit_start) = :sessionDate AND " +
+      "(v.visit_start >= :sessionDate AND v.visit_start < CAST(:sessionDate AS DATE) + CAST('1 day' AS INTERVAL)) AND " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:128]: Also the sessionStartTime and endtime check might not be needed anymore.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt [hunk 3]
```diff
@@ -125,7 +125,7 @@ interface VisitRepository : JpaRepository<Visit, Long>, JpaSpecificationExecutor
...
       "st.start_time = :sessionStartTime AND " +
       "st.end_time = :sessionEndTime AND " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:130]: Also the sessionStartTime and endtime check might not be needed anymore.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt (hunk 1 above)
Comment (rj-adams) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:107]: Would be good to see what the query planner does with this query - might point us towards ways we can improve it further

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt:130]:
```
"st.start_time = :sessionStartTime AND " +
"st.end_time = :sessionEndTime AND " +
```
Is key to the logic I dont understand why we would remove it?

Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitRepository.kt]: Your right, all I need is the session template ref

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Pull request merged**